src/Features/Core/Portable/MetadataAsSource/FileSystemMetadataDocumentPersister.cs

```diff
@@ -0,0 +1,177 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MetadataAsSource;
+
+internal class FileSystemMetadataDocumentPersister : IMetadataDocumentPersister
+{
+    /// <summary>
+    /// We create a mutex so other processes can see if our directory is still alive.  As long as we own the mutex, no
+    /// other VS instance will try to delete our _rootTemporaryPathWithGuid folder.
+    /// </summary>
+#pragma warning disable IDE0052 // Remove unread private members
+    private readonly Mutex _mutex;
+#pragma warning restore IDE0052 // Remove unread private members
+    private readonly string _rootTemporaryPathWithGuid;
+    private readonly string _rootTemporaryPath = Path.Combine(Path.GetTempPath(), MetadataAsSourceFileService.MetadataAsSource);
+
+    public FileSystemMetadataDocumentPersister()
+    {
+        var guidString = Guid.NewGuid().ToString("N");
+        _rootTemporaryPathWithGuid = Path.Combine(_rootTemporaryPath, guidString);
+        _mutex = new Mutex(initiallyOwned: true, name: CreateMutexName(guidString));
+    }
+
+    private static string CreateMutexName(string directoryName)
+        => $"{MetadataAsSourceFileService.MetadataAsSource}-{directoryName}";
+
+    public Task<SourceText?> TryGetExistingTextAsync(string documentPath, Encoding encoding, SourceHashAlgorithm checksumAlgorithm, Func<SourceText, bool> verifyExistingDocument, CancellationToken cancellationToken)
+    {
+        return TryGetExistingTextFromDisk(documentPath, encoding, checksumAlgorithm, verifyExistingDocument) is SourceText sourceText
+            ? Task.FromResult<SourceText?>(sourceText)
+            : SpecializedTasks.Null<SourceText>();
+    }
```

Comment:
the logic of this method seems odd to me.  it says TryGetExistingTextFromDisk which i would expect would be async, since it is explicitly doing IO.  but it's actually sync, but then wrapped in this async function.